node/src/components/consensus/highway_core/state.rs

```diff
@@ -351,20 +346,42 @@ impl<C: Context> State<C> {
         self.panorama[idx] = Observation::Faulty;
     }
 
+    /// Add set of endorsements to the state.
+    /// If, after adding, we have collected enough endorsements to consider vote _endorsed_,
+    /// it will be *upgraded* to fully endorsed.
     pub(crate) fn add_endorsements(&mut self, endorsements: Endorsements<C>) {
         let vote = *endorsements.vote();
         let validator_count = self.validator_count();
         info!("Received endorsements of {:?}", vote);
-        let entry = self
-            .endorsement
-            .entry(vote)
-            .or_insert_with(|| Vec::with_capacity(validator_count));
-        for (vid, signature) in endorsements.endorsers {
-            if !entry.iter().any(|e| e.validator_idx() == vid) {
-                let endorsement = Endorsement::new(vote, vid, signature);
-                entry.push(endorsement)
+        {
+            let entry = self
+                .incomplete_endorsements
+                .entry(vote)
+                .or_insert_with(|| Vec::with_capacity(validator_count));
+            for (vid, signature) in endorsements.endorsers {
+                // Add endorsements from validators we haven't seen endorsement yet.
+                if !entry.iter().any(|e| e.validator_idx() == vid) {
+                    let endorsement = Endorsement::new(vote, vid, signature);
+                    entry.push(endorsement)
+                }
             }
         }
+        // Stake required to consider vote to be endorsed.
+        let threshold = self.total_weight() / 3 * 2;
+        let endorsed = self
+            .incomplete_endorsements
+            .get(&vote)
+            .unwrap()
+            .iter()
+            .fold(Weight(0), |acc, e| {
+                let v_id = e.validator_idx();
+                acc + self.weight(v_id)
+            });
```

Comment:
[`sum()` should work, too.](https://github.com/CasperLabs/casper-node/blob/9520a8a069985760ed6904a1f2bba2b0e7056dbd/node/src/components/consensus/highway_core/state/weight.rs#L34)

Reply:
Done.